### ISSUE TITLE
Why not using worker'CPU as PS?

### ISSUE THREAD

**Issue by Shuming19** [bytedance/byteps#60]
hello:

May I ask why not using workers' CPUs as PS?

Why should PS be placed on a physically different machine from workers? There should be data transfer through worker and PS.

If local CPU is used as PS, the cost between local server and worker should be less.

Can you help me understand this point?

**Comment by ymjiang** [comment]
The key idea is that co-locating servers and workers causes bandwidth contention. If you already read [this](https://github.com/bytedance/byteps/blob/master/docs/rationale.md#the-ps-communication-pattern-is-better-theoretically), let me give you an example. 

Imagine you have 4 physical machines, and put one worker and one server on each physical machine (e.g., worker0 and server0 on same machine0). Say the model size is N. In each iteration, worker0 needs to sends out N/4 data to server1, server2 and server3, respectively (in total 3N/4); meanwhile, server0 needs to sends out N/4 data to worker1, worker2 and worker3, respectively (in total 3N/4). Therefore, machine0 needs to send out 3N/2 data in one iteration. You will find this value is equivalent to allreduce case.

In short, if you co-locate workers & servers, the communication overhead is the same as allreduce. Yes, the cost between local server and worker should be less, but it causes higher networking traffic.

**Comment by Shuming19** [comment]
In the same example, if a remote server is used, a remote server(totally 4 servers and 4 workers) needs to receive M/4 data from each worker and send M/4 data to each worker. The total data worker 0 transfers is 2M, which is greater than 3M/2, isn't it? That is to say, the server's bandwidth bottleneck is more  serious.

Expect your reply.  Thanks.

**Comment by ymjiang** [comment]
I am talking about one direction. Sending and receiving are duplex (two different directions) and do not affect each other in terms of bandwidth.

**Comment by Shuming19** [comment]
Thank you. That makes sense.

**Comment by ymjiang** [comment]
Closing this.
